Lib/Src/CUartCom.cpp

```diff
@@ -10,96 +10,239 @@
+        b_success = false;
+        return b_success;
+    }
+    else
     {
-        // Hang application with a toggling ALARM Pin
-        while (1)
+        mp_huart = p_huart;
+        m_uart_de_pin.init(uart_de_port, uart_de_pin);
+    }
+    // Check this hardware is not assigned to another instance
+    for (uint8_t i = 0; i < s_uart_instances; i++)
+    {
+        if (p_huart->Instance == sp_UART[i]->mp_huart->Instance)
         {
-            // TODO: cannot hardcode stuff like this into classes.
-            //            HAL_GPIO_TogglePin(ALARM_GPIO_Port, ALARM_Pin);
-            HAL_Delay(200);
+            b_success = false;
+            return b_success;
         }
     }
+
+    // Assign hardware to instance
+    sp_UART[s_uart_instances] = this;
+    s_uart_instances++;
+
+    return b_success;
 }
 
 /**
- * @brief sends message via UART.
+ * @brief Enables UART reception with interrupts
+ * @note UART Interrupts must be enabled in the hardware configuration in order
+ * to use this method
+ *
+ */
+void CUartCom::startRx()
+{
+    HAL_UART_Receive_IT(mp_huart, &m_rx_char, 1);
+}
+
+/**
+ * @brief Start UART Transmission and push message to the TX Queue
  * @param msg Message to send.
+ * TODO: Add return value (True if there was space in the queue)
  */
-void CUartCom::sendMessage(const std::string &msg)
+void CUartCom::send(const std::string msg)
 {
-    uint16_t msg_len = msg.length();
+    bool event = NO_MESSAGE_AVAILABLE;
 
-    // Enable UART_DE Pin
-    if (m_uart_de_port != nullptr)
+    if (m_tx_queue.size() <= MAX_TX_QUEUE_SIZE && (msg.empty() == false))
     {
-        HAL_GPIO_WritePin(m_uart_de_port, m_uart_de_pin, GPIO_PIN_SET);
+        m_tx_queue.push(msg);
     }
 
-    // Convert string to C style
-    const char *c_msg = msg.c_str();
+    if (m_tx_queue.empty() == false)
+    {
+        event = MESSAGE_AVAILABLE;
+    }
 
-    // Check it is a valid string and send error message if not
-    if (!(*c_msg) || c_msg == nullptr)
+    if (m_status == IDLE)
     {
-        const uint8_t error_msg_len = 28;
-        const char *error_msg = "Invalid string from c_str()\n";
-        HAL_UART_Transmit(mp_huart,
-                          (uint8_t *)error_msg,
-                          error_msg_len,
-                          UART_TIMEOUT);
-
-        // Disable UART_DE Pin
-        if (m_uart_de_port != nullptr)
+        if (event == NO_MESSAGE_AVAILABLE)
+        {
+            endTx();
+        }
+        else if (event == MESSAGE_AVAILABLE)
         {
-            HAL_GPIO_WritePin(m_uart_de_port, m_uart_de_pin, GPIO_PIN_RESET);
+            updateTxBuffer();
+            transmit();
         }
+    }
+}
+
+/**
+ * @brief Update TX Buffer with next element on queue
+ */
+void CUartCom::updateTxBuffer()
+{
+    std::string message = m_tx_queue.front();
+    m_tx_msg_length = message.length();
+    strcpy(m_tx_buffer, message.c_str());
+    m_status = TX;
+    m_tx_queue.pop();
+}
+
+void CUartCom::transmit()
```

Comment:
know this is a short method, but I'll keep it in case I need to disable RX Interrupts when transmitting.

Reply:
Short methods are good.